### ISSUE TITLE
Can't use DSL in standard logback.groovy 

### ISSUE THREAD

**Issue by brdbry** [virtualdogbert/logback-groovy-config#8]
**Edit**: I've just seen this issue, which I assume is what I'm seeing: https://github.com/virtualdogbert/logback-groovy-config/issues/7#issue-1705613222

Hello, thanks for picking up logback groovy support.

I'm having a few problems with the compilation of my logback-config.groovy file.

Per the docs, I've gone through adding all the imports and static imports to my logbackCompiler.groovy (including the ones in DefaultAllowLists, which didn't seem to get picked up automatically). This was a bit cumbersome (it only tells you one compilation error per startup), but fine. But I've now hit an issue whereby standard logback DSL methods are being blocked e.g.

`rollingPolicy(TimeBasedRollingPolicy){...`

Even if I knew how to whitelist a MethodCallExpression (which I don't), I surely shouldn't have to do this for the whole DSL. Any thoughts on what's going wrong here?

Thanks in advance.

<img width="903" alt="image" src="https://github.com/virtualdogbert/logback-groovy-config/assets/51187271/8c288650-529f-451b-b7c0-335a96099b57">

**Comment by virtualdogbert** [comment]
I am aware of the issue, and I do have a fix for it. I just have to finish it up and release it at some point today.

This part of the security does not and will not have an external way to whitelist, so if but more likely when, more instances are found, a bug like this will need to be raised. 

I inherited the DSL but haven't really had the time to dive deep into it. The security part which I added does have a white list for calls on Object, String, and System, because those are the most problematic when it comes to security. The DSL's calls, because of Groovy's nature, all end up on Object.

So far the list I have for Object is this:
```

    private static final List<String> AllowedObjectMethods = [
            'clone', 'equals', 'toString',
            'any', 'asBoolean', 'collect',
            'contains', 'count', 'each',
            'eachWithIndex', 'equals', 'every',
            'find', 'findIndexOf', 'findIndexValues',
            'findLastIndexOf', 'findResult', 'flatten',
            'getAt', 'grep', 'groupBy',
            'inject', 'is', 'join',
            'putAt', 'size', 'sum',
            'with',


            'conversionRule', 'appender', 'encoder',
            'forName', 'isDevelopmentMode', 'logger',
            'rollingPolicy', 'triggeringPolicy',
            'root', 'layout', 'appenderRef',
            'putProperty', 'getProperty', 'filter',
            'evaluator','println'
    ]
```

If you know of any others, offhand that should be on this list let me know. Thanks.

**Comment by virtualdogbert** [comment]
Just finished the 1.14.2,1.13.2, and 1.12.2 releases which should fix the issue with rollingPolicy. I'm not sure how long that takes to get synced though. Let me know if you find any other issues. Thanks

**Comment by brdbry** [comment]
@virtualdogbert Thanks.
After bumping to 1.12.2 the `MethodCallException` is gone but I still got `Importing [java.lang.Object.rollingPolicy] is not allowed`. For some reason it seems the `importsAcceptList` and `staticImportsAcceptList` in your `DefaultAcceptList` aren't being picked up. This was also the case with 1.12.1. May be worth creating a separate issue for that.
Anyway, as before, I copied them into my `logbackComplier.groovy`, and my logging is now working - thanks for your help.

**Comment by virtualdogbert** [comment]
Strange because I'm using the 1.12.2 in a Grails 5 upgrade that has a rollingPolicy with no issue. Can you share an example of your logback-config.groovy, the logbackComplier.groovy, and a copy of the error you're seeing? I haven't seen that the import wasn't allowed but I did have to add the import, along with others.

**Comment by virtualdogbert** [comment]
So if you find you still have this issue or someone else finds this issue they can reopen it and provide a sample app with sample config to reproduce the issue and I'll look at it.

So for now I'm closing this issue.

**Comment by brdbry** [comment]
Apologies, the application with this issue is air-gapped so I couldn't submit a sample

**Comment by virtualdogbert** [comment]
No worries I think I can see a few things missing from the original post, TimeBasedRollingPolicy, ExcludeValueOnlyFilter, FileNamePattern. I'll look into adding those in a future release. Having a sample of the logback config would be helpful so I don't have to play whack-a-mole so much, although that's why I provided the override because there is a limit to how much of that I will do. Also, there is a limit on how much free time I have, and there are always all those other side projects I want to work on...

**Comment by brdbry** [comment]
Sorry for the delay. Below are my `logback-config.groovy` and `logbackCompiler.groovy` files, which are now working in my app. Things in the `leaf` package are my app code (e.g. custom filters) so don't need to be added in the library.

As I mentioned above (https://github.com/virtualdogbert/logback-groovy-config/issues/8#issuecomment-1548143226), the odd thing is that, irrespective of whether you have specified items in the library's `DefaultAcceptLists.groovy` file, I seem to have to re-specify all of them in my `logbackCompiler.groovy` file. And one other thing: I couldn't seem to get `statusListener(OnConsoleStatusListener)` working, even using `logbackCompiler.groovy`.

I can't easily send you a sample app demonstrating this, but hope this helps.


logback-config.groovy:

```
import ch.qos.logback.core.FileAppender
import ch.qos.logback.core.rolling.RollingFileAppender
import ch.qos.logback.core.rolling.TimeBasedRollingPolicy
import ch.qos.logback.classic.encoder.PatternLayoutEncoder
import ch.qos.logback.classic.filter.ThresholdFilter
import org.springframework.boot.logging.logback.ColorConverter
import org.springframework.boot.logging.logback.WhitespaceThrowableProxyConverter
import grails.util.Environment
import leaf.lib.EnvUtils
import leaf.logging.ExcludeValOnlyAndPerfLogFilter
import leaf.logging.OnlyValOnlyLogFilter
import leaf.logging.OnlyPerformanceLogFilter
import java.nio.charset.Charset
import ch.qos.logback.core.ConsoleAppender
//import ch.qos.logback.core.status.OnConsoleStatusListener

// From logback docs: We highly recommended that you always add a status listener just
// after the last import statement and before all other statements
// Even after adding to logbackCompiler.groovy we get this: java.lang.SecurityException: Expression [MethodCallExpression] is not allowed: this.statusListener(ch.qos.logback.core.status.OnConsoleStatusListener)
//statusListener(OnConsoleStatusListener)

conversionRule 'clr', ColorConverter
conversionRule 'wex', WhitespaceThrowableProxyConverter

String performanceLogPattern = '%d{yyyy-MM-dd HH:mm:ss.SSS},' +
        '%t,' +
        '%X{requestId},' +
        '%X{userLogon},' +
        '%logger,' +
        'valOnly:%X{valOnly},' +
        '%m%n'

//With colours
String leafPatternConsole = '%clr(%d{yyyy-MM-dd HH:mm:ss.SSS}){faint} ' + // Date
        '%clr(%5p) ' + // Log level
        '%clr(---){faint} %clr([%15.15t]){faint} ' + // Thread
        '%-26X{requestId} ' +
        '%-10X{userLogon} ' +
        '%clr(%-60.60logger{59}){cyan} %clr(:){faint} ' +
        '%m%n%wex' // Message
//Without colours
String leafPatternFile = '%d{yyyy-MM-dd HH:mm:ss.SSS} ' + // Date
        '%5p ' + // Log level
        '--- [%15.15t] ' + // Thread
        '%-26X{requestId} ' + //Request
        '%-10X{userLogon} ' + //User
        '%-60.60logger{59} : ' + // Logger
        '%m%n%wex' // Message


// See http://logback.qos.ch/manual/groovy.html for details on configuration
appender('STDOUT', ConsoleAppender) {
    filter(ExcludeValOnlyAndPerfLogFilter) //exclude valOnly logs
    encoder(PatternLayoutEncoder) {
        charset = Charset.forName('UTF-8')
        pattern = leafPatternConsole
    }
}
String osNam = System.getProperty("os.name")
def ROOT_VOL = osNam.contains("Windows") ? "D:" : "/local"
def logDir = "${ROOT_VOL}/Leaf/leaf-${EnvUtils.leafEnv()}/Logs"
if(logDir == null || logDir == ''){
    def errMsg = 'Log folder not configured in Logback.groovy'
    throw new Exception(errMsg)
}

def logNamePrefix = Environment.isDevelopmentMode() ? System.getProperty('user.name') : "service"

appender('FILE', RollingFileAppender) {
    filter(ExcludeValOnlyAndPerfLogFilter) //exclude valOnly logs
    encoder(PatternLayoutEncoder) {
        charset = Charset.forName('UTF-8')
        pattern = leafPatternFile
    }
    rollingPolicy(TimeBasedRollingPolicy) {
        FileNamePattern = "${logDir}/Leaf.${logNamePrefix}.%d{yyyy-MM-dd}.log"
    }
}

appender('FILE_VALONLY', RollingFileAppender) {
    filter(OnlyValOnlyLogFilter) //valOnly logs
    encoder(PatternLayoutEncoder) {
        charset = Charset.forName('UTF-8')
        pattern = leafPatternFile
    }
    rollingPolicy(TimeBasedRollingPolicy) {
        FileNamePattern = "${logDir}/Leaf.valOnly.${logNamePrefix}.%d{yyyy-MM-dd}.log"
    }
}

appender('FILE_PERFORMANCE', RollingFileAppender) {
    filter(OnlyPerformanceLogFilter)
    encoder(PatternLayoutEncoder) {
        charset = Charset.forName('UTF-8')
        pattern = performanceLogPattern
    }
    rollingPolicy(TimeBasedRollingPolicy) {
        FileNamePattern = "${logDir}/Leaf.performance.${logNamePrefix}.%d{yyyy-MM-dd}.log"
    }
}

appender('FILE_ERROR', RollingFileAppender) {
    filter(ExcludeValOnlyAndPerfLogFilter) //exclude valOnly logs
    filter(ThresholdFilter){
        level = 'ERROR'
    }
    encoder(PatternLayoutEncoder) {
        charset = Charset.forName('UTF-8')
        pattern = leafPatternFile
    }
    rollingPolicy(TimeBasedRollingPolicy) {
        FileNamePattern = "${logDir}/Leaf.error.${logNamePrefix}.%d{yyyy-MM-dd}.log"
    }
}

logger('leaf', DEBUG)
logger('org.hibernate.engine.jdbc.spi.SqlExceptionHelper', DEBUG) //errors on hibernate SQL
logger('org.hibernate.orm.deprecation', OFF)

if (Environment.isDevelopmentMode() && logDir != null) {
    appender("FULL_STACKTRACE", FileAppender) {
        file = "${logDir}/stacktrace.log"
        append = true
        encoder(PatternLayoutEncoder) {
            pattern = "%level %logger - %msg%n"
        }
    }
    logger('leaf', TRACE)
    logger("StackTrace", ERROR, ['FULL_STACKTRACE'], false)
}

root(INFO, ['STDOUT','FILE','FILE_VALONLY', 'FILE_PERFORMANCE', 'FILE_ERROR'])
```

logbackCompiler.groovy:

```
//Per lib docs, any custom imports requred in logback.groovy should be set here
//See: https://virtualdogbert.github.io/logback-groovy-config/
importsAcceptList = [
        'ch.qos.logback.classic.filter.ThresholdFilter',
        'ch.qos.logback.classic.encoder.PatternLayoutEncoder',
        'grails.util.Environment',
        'leaf.lib.EnvUtils',
        'leaf.logging.ExcludeValOnlyLogFilter',
        'leaf.logging.OnlyValOnlyLogFilter',
        'leaf.logging.OnlyPerformanceLogFilter',
        'leaf.logging.ExcludeValOnlyAndPerfLogFilter',
        'org.springframework.boot.logging.logback.ColorConverter',
        'org.springframework.boot.logging.logback.WhitespaceThrowableProxyConverter',
        'java.nio.charset.Charset',
        'java.lang.String',
        'java.lang.Exception',

        //per docs these should be built in to DefaultAccepLists, but for some reason it was still complaining so copy them here:
        'ch.qos.logback.core.testUtil.StringListAppender',
        'java.lang.Object',
        'org.springframework.beans.factory.annotation.Autowired', //Grails requires this for some reason, but you can not autowire any service because those classes are not on the import list.
        'java.nio.charset.Charset.forName',
        'com.logentries.logback.LogentriesAppender',
        'grails.util.BuildSettings',
        'grails.util.Environment',
        'org.slf4j.MDC',
        'org.springframework.boot.logging.logback.ColorConverter',
        'org.springframework.boot.logging.logback.WhitespaceThrowableProxyConverter',
        'java.nio.charset.Charset',
        'java.nio.charset.StandardCharsets',

        'ch.qos.logback.classic.AsyncAppender',
        'ch.qos.logback.classic.boolex.JaninoEventEvaluator',
        'ch.qos.logback.classic.encoder.PatternLayoutEncoder',
        'ch.qos.logback.classic.Level',
        'ch.qos.logback.core.ConsoleAppender',
        'ch.qos.logback.core.status.OnConsoleStatusListener',
        'ch.qos.logback.core.FileAppender',
        'ch.qos.logback.classic.PatternLayout',
        'ch.qos.logback.core.encoder.LayoutWrappingEncoder',
        'ch.qos.logback.classic.LoggerContext',
        'ch.qos.logback.core.boolex.Matcher',
        'ch.qos.logback.core.filter.EvaluatorFilter',
        'java.lang.System',
        'java.lang.System.getenv',
        'java.lang.System.getProperty',
        'java.lang.Object.getOrDefault',
        'java.lang.Object.getOrDefault',

        'ch.qos.logback.core.rolling.TriggeringPolicyBase',
        'ch.qos.logback.core.rolling.helper.Compressor',
        'ch.qos.logback.core.rolling.helper.PeriodicityType',
        'ch.qos.logback.core.rolling.helper.TokenConverter',
        'ch.qos.logback.core.rolling.helper.IntegerTokenConverter',
        'ch.qos.logback.core.rolling.helper.CompressionMode',
        'ch.qos.logback.core.rolling.helper.ArchiveRemover',
        'ch.qos.logback.core.rolling.helper.FileFilterUtil',
        'ch.qos.logback.core.rolling.helper.RenameUtil',
        'ch.qos.logback.core.rolling.helper.DateTokenConverter',
        'ch.qos.logback.core.rolling.helper.FileNamePattern',
        'ch.qos.logback.core.rolling.helper.RollingCalendar',
        'ch.qos.logback.core.rolling.helper.FileStoreUtil',
        'ch.qos.logback.core.rolling.helper.SizeAndTimeBasedArchiveRemover',
        'ch.qos.logback.core.rolling.helper.TimeBasedArchiveRemover',
        'ch.qos.logback.core.rolling.helper.MonoTypedConverter',
        'ch.qos.logback.core.rolling.RollingPolicyBase',
        'ch.qos.logback.core.rolling.RollingFileAppender',
        'ch.qos.logback.core.rolling.FixedWindowRollingPolicy',
        'ch.qos.logback.core.rolling.TimeBasedFileNamingAndTriggeringPolicyBase',
        'ch.qos.logback.core.rolling.TimeBasedFileNamingAndTriggeringPolicy',
        'ch.qos.logback.core.rolling.SizeAndTimeBasedRollingPolicy',
        'ch.qos.logback.core.rolling.RollingPolicy',
        'ch.qos.logback.core.rolling.TimeBasedRollingPolicy',
        'ch.qos.logback.core.rolling.DefaultTimeBasedFileNamingAndTriggeringPolicy',
        'ch.qos.logback.core.rolling.SizeBasedTriggeringPolicy',
        'ch.qos.logback.core.rolling.RolloverFailure',
        'ch.qos.logback.core.rolling.SizeAndTimeBasedFNATP',
        'ch.qos.logback.core.rolling.TriggeringPolicy',
        'ch.qos.logback.classic.net.SMTPAppender'
]
staticImportsAcceptList = [
        'java.lang.String.contains',
        'leaf.lib.EnvUtils.leafEnv',

        'java.lang.Object.statusListener',

        //per library docs these should be built in to DefaultAccepLists, but for some reason it was still complaining so copied them here:
        'java.nio.charset.Charset.forName',
        'java.lang.Object.conversionRule',
        'java.lang.Object.appender',
        'java.lang.Object.layout',
        'java.lang.Object.appenderRef',
        'java.lang.Object.encoder',
        'java.lang.Object.filter',
        'java.lang.Object.evaluator',
        'java.lang.Object.logger',
        'java.lang.Object.root',
        'ch.qos.logback.classic.Level',
        'ch.qos.logback.classic.Level.OFF',
        'ch.qos.logback.classic.Level.ERROR',
        'ch.qos.logback.classic.Level.WARN',
        'ch.qos.logback.classic.Level.INFO',
        'ch.qos.logback.classic.Level.DEBUG',
        'ch.qos.logback.classic.Level.TRACE',
        'ch.qos.logback.classic.Level.ALL',
        'ch.qos.logback.core.spi.FilterReply',
        'ch.qos.logback.core.spi.FilterReply.DENY',
        'ch.qos.logback.core.spi.FilterReply.NEUTRAL',
        'ch.qos.logback.core.spi.FilterReply.ACCEPT',
        'ch.qos.logback.core.boolex.Matcher.start',
        'java.lang.Object.putProperty',
        'java.lang.Object.getProperty',
        'java.lang.System',
        'java.lang.System.getenv',
        'java.lang.System.getProperty',
        'java.lang.Object.getOrDefault',
        'java.lang.Object.println',
        'java.lang.Object.rollingPolicy',
        'java.lang.Object.triggeringPolicy'

]
```

**Comment by virtualdogbert** [comment]
No worries, I'm just a software engineer(software "architect"), I imagine not unlike you. With my various open-source projects, I only get to them when I can make the time.

This is good and should allow me to fix the issues you are having because now I understand what you have and what the issues are. statusListener would be a DSL method so I will have to add that to the ScriptExpressionChecker. This will not be made configurable outside of the library for security reasons. For the list of DSL methods, I wasn't given a list so I had to search through the code and hope I found all o them. Because of that, I expected I would get some bug reports like this.

As for the imports in logbackCompiler.groovy, right now for the overrides, well override, they are not additive, but I could change that to make, the library a bit easier to use.

I'll get on these changes next week once I'm back from vacation. I'll also put together a shell app with some instructions on how to make a sample app to try to make interactions like this a little easier. Mostly what I needed to see was the logback-config.groovy and the logbackCompiler.groovy. So what I'm thinking in the future people submitting bugs could fork the sample app, drop in their config files, and set/add dependencies.

**Comment by virtualdogbert** [comment]
Just released new versions 1.14.4, 1.13.4, and 1.12.4, based on your logback-config.groovy. For what you have you would still have to specify a logbackCompiler.groovy, but now, in theory, you should only have to add `leaf.logging.*` entries, as those are not part of logback.

It may take a while for the libraries to show up on search websites for maven central, but they should be accessible sometime later today. It usually takes a little while for them to be accessible.

**Comment by brdbry** [comment]
Sorry for the delay. Thanks for the fix. However for me it didn't allow me to tidy up my logbackCompiler.groovy - I still got a security error on import of `java.lang.String`. Just going to leave it as is :-) Thanks again.
